rllib/offline/estimators/direct_method.py

```diff
@@ -0,0 +1,119 @@
+from typing import Dict
+from ray.rllib.offline.estimators.off_policy_estimator import (
+    OffPolicyEstimator,
+    OffPolicyEstimate,
+)
+from ray.rllib.policy import Policy
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.numpy import convert_to_numpy
+from ray.rllib.utils.typing import SampleBatchType
+from ray.rllib.offline.estimators.qreg_tf_model import QRegTFModel
+from ray.rllib.offline.estimators.qreg_torch_model import QRegTorchModel
+from gym.spaces import Discrete
+import numpy as np
+
+
+def k_fold_cv(batch: SampleBatchType, k: int):
+    """Utility function that returns a k-fold cross validation generator
+    over episodes from the given batch.
+
+    Args:
+        batch: A SampleBatch of episodes to split
+        k: Number of cross-validation splits
+
+    Returns:
+        A tuple of SampleBatches (train_episodes, test_episodes)
+    """
+    episodes = batch.split_by_episode()
+    n_episodes = len(episodes)
+    if n_episodes < k:
+        # TODO(rohan): print warning: "len(batch) < k, running OPE without training"
+        yield [], episodes
+        return
+    n_fold = n_episodes // k
+    for i in range(k):
+        train_episodes = episodes[: i * n_fold] + episodes[(i + 1) * n_fold :]
+        if i != k - 1:
+            test_episodes = episodes[i * n_fold : (i + 1) * n_fold]
+        else:
+            # Append remaining episodes onto the last test_episodes
+            test_episodes = episodes[i * n_fold :]
+        yield train_episodes, test_episodes
+    return
+
+
+class DirectMethod(OffPolicyEstimator):
+    """The Direct Method (Q-Reg) estimator.
+
+    config: {
+        model: ModelConfigDict,
+        k: k-fold cross validation for training model and evaluating OPE
+    }
+
+    Q-Reg estimator described in https://arxiv.org/pdf/1511.03722.pdf,
+    https://arxiv.org/pdf/1911.06854.pdf"""
+
+    @override(OffPolicyEstimator)
+    def __init__(self, policy: Policy, gamma: float, config: Dict):
+        super().__init__(policy, gamma, config)
+        assert isinstance(
+            policy.action_space, Discrete
+        ), "DM Estimator only supports discrete action spaces!"
+        assert (
+            policy.config["batch_mode"] == "complete_episodes"
+        ), "DM Estimator only supports batch_mode=`complete_episodes`"
+        model_cls = QRegTorchModel if policy.framework == "torch" else QRegTFModel
+        self.model = model_cls(
+            policy=policy,
+            gamma=gamma,
+            config=config,
+        )
+        self.k = config.get("k", 5)
+
+    @override(OffPolicyEstimator)
+    def estimate(self, batch: SampleBatchType) -> OffPolicyEstimate:
+        self.check_can_estimate_for(batch)
+        estimates = []
+        # Split data into train and test using k-fold cross validation
```

Comment:
as we discussed offline, can we add a TODO here saying that potentially, this can be handled by AIR cross-validation?
Then the Direct estimator here can be a simple 1-time training run.